packages/actor-rdf-join-multi-smallest/lib/ActorRdfJoinMultiSmallest.ts

```diff
@@ -0,0 +1,63 @@
+import {IActorQueryOperationOutput, IActorQueryOperationOutputBindings} from "@comunica/bus-query-operation";
+import {ActorRdfJoin, IActionRdfJoin} from "@comunica/bus-rdf-join";
+import {IActorArgs, IActorTest, Mediator} from "@comunica/core";
+import {IMediatorTypeIterations} from "@comunica/mediatortype-iterations";
+
+/**
+ * A comunica Multi Smallest RDF Join Actor.
+ * It accepts 3 or streams at once, joins the smallest two, and joins the result with the remaining streams.
+ */
+export class ActorRdfJoinMultiSmallest extends ActorRdfJoin {
+
+  public readonly mediatorJoin: Mediator<ActorRdfJoin,
+    IActionRdfJoin, IMediatorTypeIterations, IActorQueryOperationOutput>;
+
+  constructor(args: IActorRdfJoinMultiSmallestArgs) {
+    super(args, 3, true);
+  }
+
+  public static getSmallestPatternId(totalItems: number[]) {
+    let smallestId: number = -1;
+    let smallestCount: number = Infinity;
+    for (let i = 0; i < totalItems.length; i++) {
+      const count: number = totalItems[i];
+      if (count <= smallestCount) {
+        smallestCount = count;
+        smallestId = i;
+      }
+    }
+    return smallestId;
+  }
+
+  protected async getOutput(action: IActionRdfJoin): Promise<IActorQueryOperationOutputBindings> {
+    const entries: IActorQueryOperationOutputBindings[] = action.entries.slice();
+
+    // Determine the two smallest streams by estimated count
+    const entriesTotalItems = (await Promise.all(action.entries.map((entry) => entry.metadata())))
+      .map((metadata) => 'totalItems' in metadata ? metadata.totalItems : Infinity);
+    const smallestIndex1: number = ActorRdfJoinMultiSmallest.getSmallestPatternId(entriesTotalItems);
```

Comment:
it's only going to be relevant once there are many many streams that need to be joined at the same time, but why not make the function getTwoSmallestPatternIds?

Reply:
I think the implementation would become too complex and unreadable, but feel free to give it ago if you think you can do it in a consise way :-)

Reply:
I wrote it in notepad so there might be syntax errors, but something like this

```javascript
  public static getSmallestPatternIds(totalItems: number[], count) {
    const smallest = [];
    for (let j = 0; j < count; ++j) {
      smallest.push({ id: -1, count: Infinity });
    }
    
    for (let i = 0; i < totalItems.length; i++) {
      const count: number = totalItems[i];
      for (let k = count-1; k >= 0; --k) {
        // insert new entry if it is smaller
        if (count < smallest[k].count) {
            smallest[k].id = i;
            smallest[k].count = count;
        }
        // stop if the next entry would be larger (or we reached the end)
        if (count >= smallest[k].count || k === 0) {
          // shift larger entries to the right
          if (k < count-1) {
            smallest[k+1] = smallest[k];
          }
          break;
        }
      }
    }
    return smallest;
  }
```
I'll leave it up to you if it's concise enough :P

Reply:
I've tried implementing it here: https://github.com/comunica/comunica/tree/feature/multi-join-abstracter
It doesn't seem to be working correctly though. Feel free to debug it further if you're up for it. I'll merge this PR in the meantime. We can still apply your improved algorithm at a later stage.